File: oxidb-spring-boot-starter/src/main/java/com/oxidb/spring/OxiDbClient.java
package com.oxidb.spring;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;
import java.util.Map;

/**
 * TCP client for oxidb-server.
 * <p>
 * Protocol: each message is [4-byte little-endian length][JSON payload].
 * Server responds with {"ok": true, "data": ...} or {"ok": false, "error": "..."}.
 * <p>
 * Thread-safe: all send/receive operations are synchronized.
 * Implements AutoCloseable for try-with-resources and Spring lifecycle.
 */
public class OxiDbClient implements AutoCloseable {

    private final ObjectMapper mapper = new ObjectMapper();
    private final Socket socket;
    private final DataOutputStream out;
    private final DataInputStream in;

    public OxiDbClient(String host, int port, int timeoutMs) {
        try {
            this.socket = new Socket();
            this.socket.setSoTimeout(timeoutMs);
            this.socket.connect(new InetSocketAddress(host, port), timeoutMs);
            this.out = new DataOutputStream(socket.getOutputStream());
            this.in = new DataInputStream(socket.getInputStream());
        } catch (IOException e) {
            throw new OxiDbException("Failed to connect to OxiDB at " + host + ":" + port, e);
        }
    }

    // ------------------------------------------------------------------
    // Low-level protocol
    // ------------------------------------------------------------------

    private synchronized void sendRaw(byte[] data) throws IOException {
        byte[] lenBytes = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(data.length).array();
        out.write(lenBytes);
        out.write(data);
        out.flush();
    }

    private synchronized byte[] recvRaw() throws IOException {
        byte[] lenBytes = in.readNBytes(4);
        if (lenBytes.length < 4) {
            throw new IOException("Connection closed by server");
        }
        int length = ByteBuffer.wrap(lenBytes).order(ByteOrder.LITTLE_ENDIAN).getInt();
        byte[] payload = in.readNBytes(length);
        if (payload.length < length) {
            throw new IOException("Connection closed by server");
        }
        return payload;
    }

    private synchronized JsonNode request(ObjectNode payload) {
        try {
            byte[] jsonBytes = mapper.writeValueAsBytes(payload);
            sendRaw(jsonBytes);
            byte[] respBytes = recvRaw();
            return mapper.readTree(respBytes);
        } catch (IOException e) {
            throw new OxiDbException("Communication error", e);
        }
    }

    private JsonNode checked(ObjectNode payload) {
        JsonNode resp = request(payload);
        if (!resp.path("ok").asBoolean(false)) {
            String error = resp.path("error").asText("unknown error");
            if (error.toLowerCase().contains("conflict")) {
                throw new TransactionConflictException(error);
            }
            throw new OxiDbException(error);
        }
        return resp.get("data");
    }

    private ObjectNode cmd(String command) {
        ObjectNode node = mapper.createObjectNode();
        node.put("cmd", command);
        return node;
    }

    // ------------------------------------------------------------------
    // Utility
    // ------------------------------------------------------------------

    /** Ping the server. Returns "pong". */
    public JsonNode ping() {
        return checked(cmd("ping"));
    }

    // ------------------------------------------------------------------
    // Collection management
    // ------------------------------------------------------------------

    /** Explicitly create a collection. Collections are also auto-created on insert. */
    public JsonNode createCollection(String name) {
        ObjectNode p = cmd("create_collection");
        p.put("collection", name);
        return checked(p);
    }

    /** Return a list of collection names. */
    public JsonNode listCollections() {
        return checked(cmd("list_collections"));
    }

    /** Drop a collection and its data. */
    public JsonNode dropCollection(String name) {
        ObjectNode p = cmd("drop_collection");
        p.put("collection", name);
        return checked(p);
    }

    // ------------------------------------------------------------------
    // CRUD
    // ------------------------------------------------------------------

    /** Insert a single document. Returns {"id": ...} outside tx, "buffered" inside tx. */
    public JsonNode insert(String collection, Map<String, Object> doc) {
        ObjectNode p = cmd("insert");
        p.put("collection", collection);
        p.set("doc", mapper.valueToTree(doc));
        return checked(p);
    }

    /** Insert a single document from a JSON string. */
    public JsonNode insert(String collection, String docJson) {
        try {
            ObjectNode p = cmd("insert");
            p.put("collection", collection);
            p.set("doc", mapper.readTree(docJson));
            return checked(p);
        } catch (JsonProcessingException e) {
            throw new OxiDbException("Invalid JSON: " + e.getMessage(), e);
        }
    }

    /** Insert multiple documents. */
    public JsonNode insertMany(String collection, List<Map<String, Object>> docs) {
        ObjectNode p = cmd("insert_many");
        p.put("collection", collection);
        p.set("docs", mapper.valueToTree(docs));
        return checked(p);
    }

    /** Find documents matching a query. */
    public JsonNode find(String collection, Map<String, Object> query) {
        return find(collection, query, null, null, null);
    }

    /** Find documents with sort, skip, and limit options. */
    public JsonNode find(String collection, Map<String, Object> query,
                         Map<String, Object> sort, Integer skip, Integer limit) {
        ObjectNode p = cmd("find");
        p.put("collection", collection);
        p.set("query", query != null ? mapper.valueToTree(query) : mapper.createObjectNode());
        if (sort != null) p.set("sort", mapper.valueToTree(sort));
        if (skip != null) p.put("skip", skip);
        if (limit != null) p.put("limit", limit);
        return checked(p);
    }

    /** Find documents using a JSON query string. */
    public JsonNode find(String collection, String queryJson) {
        try {
            ObjectNode p = cmd("find");
            p.put("collection", collection);
            p.set("query", mapper.readTree(queryJson));
            return checked(p);
        } catch (JsonProcessingException e) {
            throw new OxiDbException("Invalid JSON: " + e.getMessage(), e);
        }
    }

    /** Find a single document matching a query. Returns the document or null. */
    public JsonNode findOne(String collection, Map<String, Object> query) {
        ObjectNode p = cmd("find_one");
        p.put("collection", collection);
        p.set("query", query != null ? mapper.valueToTree(query) : mapper.createObjectNode());
        return checked(p);
    }

    /** Update documents matching a query. Returns {"modified": n} outside tx. */
    public JsonNode update(String collection, Map<String, Object> query, Map<String, Object> update) {
        ObjectNode p = cmd("update");
        p.put("collection", collection);
        p.set("query", mapper.valueToTree(query));
        p.set("update", mapper.valueToTree(update));
        return checked(p);
    }

    /** Update documents using JSON strings for query and update. */
    public JsonNode update(String collection, String queryJson, String updateJson) {
        try {
            ObjectNode p = cmd("update");
            p.put("collection", collection);
            p.set("query", mapper.readTree(queryJson));
            p.set("update", mapper.readTree(updateJson));
            return checked(p);
        } catch (JsonProcessingException e) {
            throw new OxiDbException("Invalid JSON: " + e.getMessage(), e);
        }
    }

    /** Delete documents matching a query. Returns {"deleted": n} outside tx. */
    public JsonNode delete(String collection, Map<String, Object> query) {
        ObjectNode p = cmd("delete");
        p.put("collection", collection);
        p.set("query", mapper.valueToTree(query));
        return checked(p);
    }

    /** Delete documents using a JSON query string. */
    public JsonNode delete(String collection, String queryJson) {
        try {
            ObjectNode p = cmd("delete");
            p.put("collection", collection);
            p.set("query", mapper.readTree(queryJson));
            return checked(p);
        } catch (JsonProcessingException e) {
            throw new OxiDbException("Invalid JSON: " + e.getMessage(), e);
        }
    }

    /** Count documents matching a query. */
    public int count(String collection, Map<String, Object> query) {
        ObjectNode p = cmd("count");
        p.put("collection", collection);
        p.set("query", query != null ? mapper.valueToTree(query) : mapper.createObjectNode());
        JsonNode result = checked(p);
        return result.path("count").asInt();
    }

    /** Count all documents in a collection. */
    public int count(String collection) {
        return count(collection, null);
    }

    // ------------------------------------------------------------------
    // Indexes
    // ------------------------------------------------------------------

    /** Create a non-unique index on a field. */
    public JsonNode createIndex(String collection, String field) {
        ObjectNode p = cmd("create_index");
        p.put("collection", collection);
        p.put("field", field);
        return checked(p);
    }

    /** Create a unique index on a field. */
    public JsonNode createUniqueIndex(String collection, String field) {
        ObjectNode p = cmd("create_unique_index");
        p.put("collection", collection);
        p.put("field", field);
        return checked(p);
    }

    /** Create a composite index on multiple fields. */
    public JsonNode createCompositeIndex(String collection, List<String> fields) {
        ObjectNode p = cmd("create_composite_index");
        p.put("collection", collection);
        p.set("fields", mapper.valueToTree(fields));
        return checked(p);
    }

    // ------------------------------------------------------------------
    // Aggregation
    // ------------------------------------------------------------------

    /** Run an aggregation pipeline. Returns list of result documents. */
    public JsonNode aggregate(String collection, List<Map<String, Object>> pipeline) {
        ObjectNode p = cmd("aggregate");
        p.put("collection", collection);
        p.set("pipeline", mapper.valueToTree(pipeline));
        return checked(p);
    }

    /** Run an aggregation pipeline from a JSON string. */
    public JsonNode aggregate(String collection, String pipelineJson) {
        try {
            ObjectNode p = cmd("aggregate");
            p.put("collection", collection);
            p.set("pipeline", mapper.readTree(pipelineJson));
            return checked(p);
        } catch (JsonProcessingException e) {
            throw new OxiDbException("Invalid JSON: " + e.getMessage(), e);
        }
    }

    // ------------------------------------------------------------------
    // Compaction
    // ------------------------------------------------------------------

    /** Compact a collection. Returns {old_size, new_size, docs_kept}. */
    public JsonNode compact(String collection) {
        ObjectNode p = cmd("compact");
        p.put("collection", collection);
        return checked(p);
    }

    // ------------------------------------------------------------------
    // Transactions
    // ------------------------------------------------------------------

    /** Begin a transaction on this connection. Returns {"tx_id": ...}. */
    public JsonNode beginTx() {
        return checked(cmd("begin_tx"));
    }

    /** Commit the active transaction. Throws TransactionConflictException on OCC conflict. */
    public JsonNode commitTx() {
        return checked(cmd("commit_tx"));
    }

    /** Rollback the active transaction. */
    public JsonNode rollbackTx() {
        return checked(cmd("rollback_tx"));
    }

    /**
     * Execute a block within a transaction. Auto-commits on success, auto-rolls back on exception.
     *
     * @param action the operations to perform within the transaction
     */
    public void withTransaction(Runnable action) {
        beginTx();
        try {
            action.run();
            commitTx();
        } catch (Exception e) {
            try {
                rollbackTx();
            } catch (OxiDbException ignored) {
                // rollback may fail if commit already failed
            }
            throw e;
        }
    }

    // ------------------------------------------------------------------
    // Blob storage
    // ------------------------------------------------------------------

    /** Create a blob storage bucket. */
    public JsonNode createBucket(String bucket) {
        ObjectNode p = cmd("create_bucket");
        p.put("bucket", bucket);
        return checked(p);
    }

    /** List all blob storage buckets. */
    public JsonNode listBuckets() {
        return checked(cmd("list_buckets"));
    }

    /** Delete a blob storage bucket. */
    public JsonNode deleteBucket(String bucket) {
        ObjectNode p = cmd("delete_bucket");
        p.put("bucket", bucket);
        return checked(p);
    }

    /** Upload a blob object. Data is base64-encoded automatically. */
    public JsonNode putObject(String bucket, String key, byte[] data,
                              String contentType, Map<String, String> metadata) {
        ObjectNode p = cmd("put_object");
        p.put("bucket", bucket);
        p.put("key", key);
        p.put("data", Base64.getEncoder().encodeToString(data));
        p.put("content_type", contentType != null ? contentType : "application/octet-stream");
        if (metadata != null && !metadata.isEmpty()) {
            p.set("metadata", mapper.valueToTree(metadata));
        }
        return checked(p);
    }

    /** Upload a blob object with default content type. */
    public JsonNode putObject(String bucket, String key, byte[] data) {
        return putObject(bucket, key, data, null, null);
    }

    /** Download a blob object. Returns the response with "content" (base64) and "metadata". */
    public JsonNode getObject(String bucket, String key) {
        ObjectNode p = cmd("get_object");
        p.put("bucket", bucket);
        p.put("key", key);
        return checked(p);
    }

    /** Decode the base64 content from a getObject response. */
    public byte[] decodeObjectContent(JsonNode getObjectResult) {
        String b64 = getObjectResult.path("content").asText();
        return Base64.getDecoder().decode(b64);
    }

    /** Get blob object metadata without downloading the content. */
    public JsonNode headObject(String bucket, String key) {
        ObjectNode p = cmd("head_object");
        p.put("bucket", bucket);
        p.put("key", key);
        return checked(p);
    }

    /** Delete a blob object. */
    public JsonNode deleteObject(String bucket, String key) {
        ObjectNode p = cmd("delete_object");
        p.put("bucket", bucket);
        p.put("key", key);
        return checked(p);
    }

    /** List objects in a bucket. */
    public JsonNode listObjects(String bucket, String prefix, Integer limit) {
        ObjectNode p = cmd("list_objects");
        p.put("bucket", bucket);
        if (prefix != null) p.put("prefix", prefix);
        if (limit != null) p.put("limit", limit);
        return checked(p);
    }

    /** List all objects in a bucket. */
    public JsonNode listObjects(String bucket) {
        return listObjects(bucket, null, null);
    }

    // ------------------------------------------------------------------
    // Full-text search
    // ------------------------------------------------------------------

    /** Full-text search across blobs. Returns [{bucket, key, score}, ...]. */
    public JsonNode search(String query, String bucket, int limit) {
        ObjectNode p = cmd("search");
        p.put("query", query);
        if (bucket != null) p.put("bucket", bucket);
        p.put("limit", limit);
        return checked(p);
    }

    /** Full-text search with default limit of 10. */
    public JsonNode search(String query) {
        return search(query, null, 10);
    }

    // ------------------------------------------------------------------
    // AutoCloseable
    // ------------------------------------------------------------------

    @Override
    public void close() {
        try {
            socket.close();
        } catch (IOException ignored) {
        }
    }
}

File: php/src/OxiDbClient.php
<?php

declare(strict_types=1);

namespace OxiDb;

/**
 * TCP client for oxidb-server.
 *
 * Protocol: each message is [4-byte little-endian length][JSON payload].
 * Server responds with {"ok": true, "data": ...} or {"ok": false, "error": "..."}.
 *
 * Zero external dependencies — uses only PHP sockets and json.
 */
class OxiDbClient
{
    /** @var resource */
    private $socket;

    /**
     * Connect to oxidb-server.
     *
     * @param string $host Server host (default "127.0.0.1")
     * @param int    $port Server port (default 4444)
     * @param float  $timeout Connect/read timeout in seconds (default 5.0)
     * @throws OxiDbException
     */
    public function __construct(string $host = '127.0.0.1', int $port = 4444, float $timeout = 5.0)
    {
        $this->socket = @fsockopen($host, $port, $errno, $errstr, $timeout);
        if ($this->socket === false) {
            throw new OxiDbException("Failed to connect to OxiDB at {$host}:{$port}: {$errstr}");
        }
        stream_set_timeout($this->socket, (int)$timeout, (int)(($timeout - (int)$timeout) * 1000000));
    }

    /**
     * Close the TCP connection.
     */
    public function close(): void
    {
        if (is_resource($this->socket)) {
            fclose($this->socket);
        }
    }

    public function __destruct()
    {
        $this->close();
    }

    // ------------------------------------------------------------------
    // Low-level protocol
    // ------------------------------------------------------------------

    private function sendRaw(string $data): void
    {
        $len = pack('V', strlen($data));
        $this->writeAll($len);
        $this->writeAll($data);
    }

    private function recvRaw(): string
    {
        $lenBytes = $this->readExact(4);
        $unpacked = unpack('Vlength', $lenBytes);
        $length = $unpacked['length'];
        return $this->readExact($length);
    }

    private function writeAll(string $data): void
    {
        $remaining = strlen($data);
        $offset = 0;
        while ($remaining > 0) {
            $written = @fwrite($this->socket, substr($data, $offset), $remaining);
            if ($written === false || $written === 0) {
                throw new OxiDbException('Failed to write to socket');
            }
            $offset += $written;
            $remaining -= $written;
        }
    }

    private function readExact(int $n): string
    {
        $buf = '';
        while (strlen($buf) < $n) {
            $chunk = @fread($this->socket, $n - strlen($buf));
            if ($chunk === false || $chunk === '') {
                throw new OxiDbException('Connection closed by server');
            }
            $buf .= $chunk;
        }
        return $buf;
    }

    /**
     * @return mixed
     * @throws OxiDbException
     */
    private function request(array $payload)
    {
        $json = json_encode($payload, JSON_THROW_ON_ERROR);
        $this->sendRaw($json);
        $respBytes = $this->recvRaw();
        return json_decode($respBytes, true, 512, JSON_THROW_ON_ERROR);
    }

    /**
     * @return mixed
     * @throws OxiDbException|TransactionConflictException
     */
    private function checked(array $payload)
    {
        $resp = $this->request($payload);
        if (empty($resp['ok'])) {
            $error = $resp['error'] ?? 'unknown error';
            if (stripos($error, 'conflict') !== false) {
                throw new TransactionConflictException($error);
            }
            throw new OxiDbException($error);
        }
        return $resp['data'] ?? null;
    }

    // ------------------------------------------------------------------
    // Utility
    // ------------------------------------------------------------------

    /**
     * Ping the server. Returns "pong".
     */
    public function ping(): string
    {
        return $this->checked(['cmd' => 'ping']);
    }

    // ------------------------------------------------------------------
    // Collection management
    // ------------------------------------------------------------------

    public function createCollection(string $name): mixed
    {
        return $this->checked(['cmd' => 'create_collection', 'collection' => $name]);
    }

    public function listCollections(): array
    {
        return $this->checked(['cmd' => 'list_collections']);
    }

    public function dropCollection(string $name): mixed
    {
        return $this->checked(['cmd' => 'drop_collection', 'collection' => $name]);
    }

    // ------------------------------------------------------------------
    // CRUD
    // ------------------------------------------------------------------

    public function insert(string $collection, array $doc): mixed
    {
        return $this->checked(['cmd' => 'insert', 'collection' => $collection, 'doc' => (object)$doc]);
    }

    public function insertMany(string $collection, array $docs): mixed
    {
        $objects = array_map(fn($d) => (object)$d, $docs);
        return $this->checked(['cmd' => 'insert_many', 'collection' => $collection, 'docs' => $objects]);
    }

    public function find(string $collection, array $query = [], ?array $sort = null, ?int $skip = null, ?int $limit = null): array
    {
        $payload = ['cmd' => 'find', 'collection' => $collection, 'query' => (object)$query];
        if ($sort !== null) $payload['sort'] = (object)$sort;
        if ($skip !== null) $payload['skip'] = $skip;
        if ($limit !== null) $payload['limit'] = $limit;
        return $this->checked($payload);
    }

    public function findOne(string $collection, array $query = []): mixed
    {
        return $this->checked(['cmd' => 'find_one', 'collection' => $collection, 'query' => (object)$query]);
    }

    public function update(string $collection, array $query, array $update): mixed
    {
        return $this->checked([
            'cmd' => 'update', 'collection' => $collection,
            'query' => (object)$query, 'update' => (object)$update,
        ]);
    }

    public function delete(string $collection, array $query): mixed
    {
        return $this->checked(['cmd' => 'delete', 'collection' => $collection, 'query' => (object)$query]);
    }

    public function count(string $collection, array $query = []): int
    {
        $result = $this->checked(['cmd' => 'count', 'collection' => $collection, 'query' => (object)$query]);
        return $result['count'];
    }

    // ------------------------------------------------------------------
    // Indexes
    // ------------------------------------------------------------------

    public function createIndex(string $collection, string $field): mixed
    {
        return $this->checked(['cmd' => 'create_index', 'collection' => $collection, 'field' => $field]);
    }

    public function createUniqueIndex(string $collection, string $field): mixed
    {
        return $this->checked(['cmd' => 'create_unique_index', 'collection' => $collection, 'field' => $field]);
    }

    public function createCompositeIndex(string $collection, array $fields): mixed
    {
        return $this->checked(['cmd' => 'create_composite_index', 'collection' => $collection, 'fields' => $fields]);
    }

    // ------------------------------------------------------------------
    // Aggregation
    // ------------------------------------------------------------------

    public function aggregate(string $collection, array $pipeline): array
    {
        return $this->checked(['cmd' => 'aggregate', 'collection' => $collection, 'pipeline' => $pipeline]);
    }

    // ------------------------------------------------------------------
    // Compaction
    // ------------------------------------------------------------------

    public function compact(string $collection): array
    {
        return $this->checked(['cmd' => 'compact', 'collection' => $collection]);
    }

    // ------------------------------------------------------------------
    // Transactions
    // ------------------------------------------------------------------

    public function beginTx(): mixed
    {
        return $this->checked(['cmd' => 'begin_tx']);
    }

    public function commitTx(): mixed
    {
        return $this->checked(['cmd' => 'commit_tx']);
    }

    public function rollbackTx(): mixed
    {
        return $this->checked(['cmd' => 'rollback_tx']);
    }

    /**
     * Execute a callable within a transaction.
     * Auto-commits on success, auto-rolls back on exception.
     *
     * @param callable $fn The operations to perform
     * @throws \Throwable
     */
    public function transaction(callable $fn): void
    {
        $this->beginTx();
        try {
            $fn();
            $this->commitTx();
        } catch (\Throwable $e) {
            try {
                $this->rollbackTx();
            } catch (OxiDbException $ignored) {
            }
            throw $e;
        }
    }

    // ------------------------------------------------------------------
    // Blob storage
    // ------------------------------------------------------------------

    public function createBucket(string $bucket): mixed
    {
        return $this->checked(['cmd' => 'create_bucket', 'bucket' => $bucket]);
    }

    public function listBuckets(): array
    {
        return $this->checked(['cmd' => 'list_buckets']);
    }

    public function deleteBucket(string $bucket): mixed
    {
        return $this->checked(['cmd' => 'delete_bucket', 'bucket' => $bucket]);
    }

    /**
     * Upload a blob object. Data is base64-encoded automatically.
     */
    public function putObject(string $bucket, string $key, string $data, string $contentType = 'application/octet-stream', ?array $metadata = null): mixed
    {
        $payload = [
            'cmd' => 'put_object',
            'bucket' => $bucket,
            'key' => $key,
            'data' => base64_encode($data),
            'content_type' => $contentType,
        ];
        if ($metadata !== null && !empty($metadata)) {
            $payload['metadata'] = (object)$metadata;
        }
        return $this->checked($payload);
    }

    /**
     * Download a blob object. Returns [data_string, metadata_array].
     */
    public function getObject(string $bucket, string $key): array
    {
        $result = $this->checked(['cmd' => 'get_object', 'bucket' => $bucket, 'key' => $key]);
        $data = base64_decode($result['content']);
        return [$data, $result['metadata']];
    }

    public function headObject(string $bucket, string $key): array
    {
        return $this->checked(['cmd' => 'head_object', 'bucket' => $bucket, 'key' => $key]);
    }

    public function deleteObject(string $bucket, string $key): mixed
    {
        return $this->checked(['cmd' => 'delete_object', 'bucket' => $bucket, 'key' => $key]);
    }

    public function listObjects(string $bucket, ?string $prefix = null, ?int $limit = null): array
    {
        $payload = ['cmd' => 'list_objects', 'bucket' => $bucket];
        if ($prefix !== null) $payload['prefix'] = $prefix;
        if ($limit !== null) $payload['limit'] = $limit;
        return $this->checked($payload);
    }

    // ------------------------------------------------------------------
    // Full-text search
    // ------------------------------------------------------------------

    public function search(string $query, ?string $bucket = null, int $limit = 10): array
    {
        $payload = ['cmd' => 'search', 'query' => $query, 'limit' => $limit];
        if ($bucket !== null) $payload['bucket'] = $bucket;
        return $this->checked($payload);
    }
}
